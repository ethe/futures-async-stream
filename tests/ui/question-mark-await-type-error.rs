// SPDX-License-Identifier: Apache-2.0 OR MIT

#![feature(generators, proc_macro_hygiene, stmt_expr_attributes)]

use futures_async_stream::{for_await, stream};

#[stream(item = i32)]
async fn stream(x: i32) {
    for i in 1..=x {
        yield i
    }
}

async fn async_fn() {
    for _i in 1..2 {
        async {}.await?; //~ ERROR the `?` operator can only be applied to values that implement `std::ops::Try`
    }
}

#[stream(item = i32)]
async fn async_stream_fn() {
    for _i in 1..2 {
        async {}.await?; //~ ERROR the `?` operator can only be applied to values that implement `std::ops::Try`
    }
}

async fn async_fn_and_for_await() {
    #[for_await]
    for _i in stream(2) {
        async {}.await?; //~ ERROR the `?` operator can only be applied to values that implement `std::ops::Try`
    }
}

#[stream(item = i32)]
async fn async_stream_fn_and_for_await() {
    #[for_await]
    for _i in stream(2) {
        async {}.await?; //~ ERROR the `?` operator can only be applied to values that implement `std::ops::Try`
    }
}

fn main() {}
