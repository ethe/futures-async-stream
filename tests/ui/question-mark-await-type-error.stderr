error[E0277]: the `?` operator can only be applied to values that implement `Try`
   --> tests/ui/question-mark-await-type-error.rs:14:9
    |
14  |         async {}.await?; //~ ERROR the `?` operator can only be applied to values that implement `std::ops::Try`
    |         ^^^^^^^^^^^^^^^ the `?` operator cannot be applied to type `()`
    |
    = help: the trait `Try` is not implemented for `()`
note: required by `branch`
   --> $RUST/core/src/ops/try_trait.rs
    |
    |     fn branch(self) -> ControlFlow<Self::Residual, Self::Output>;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the `?` operator can only be used in an async function that returns `Result` or `Option` (or another type that implements `FromResidual`)
   --> tests/ui/question-mark-await-type-error.rs:14:23
    |
12  |   async fn async_fn() {
    |  _____________________-
13  | |     for _i in 1..2 {
14  | |         async {}.await?; //~ ERROR the `?` operator can only be applied to values that implement `std::ops::Try`
    | |                       ^ cannot use the `?` operator in an async function that returns `()`
15  | |     }
16  | | }
    | |_- this function should return `Result` or `Option` to accept `?`
    |
    = help: the trait `FromResidual<_>` is not implemented for `()`
note: required by `from_residual`
   --> $RUST/core/src/ops/try_trait.rs
    |
    |     fn from_residual(residual: R) -> Self;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the `?` operator can only be applied to values that implement `Try`
   --> tests/ui/question-mark-await-type-error.rs:21:18
    |
21  |         async {}.await?; //~ ERROR the `?` operator can only be applied to values that implement `std::ops::Try`
    |                  ^^^^^^ the `?` operator cannot be applied to type `()`
    |
    = help: the trait `Try` is not implemented for `()`
note: required by `branch`
   --> $RUST/core/src/ops/try_trait.rs
    |
    |     fn branch(self) -> ControlFlow<Self::Residual, Self::Output>;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the `?` operator can only be used in a generator that returns `Result` or `Option` (or another type that implements `FromResidual`)
   --> tests/ui/question-mark-await-type-error.rs:21:23
    |
18  | #[stream(item = i32)]
    | --------------------- this function should return `Result` or `Option` to accept `?`
...
21  |         async {}.await?; //~ ERROR the `?` operator can only be applied to values that implement `std::ops::Try`
    |                       ^ cannot use the `?` operator in a generator that returns `()`
    |
    = help: the trait `FromResidual<_>` is not implemented for `()`
note: required by `from_residual`
   --> $RUST/core/src/ops/try_trait.rs
    |
    |     fn from_residual(residual: R) -> Self;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the `?` operator can only be applied to values that implement `Try`
   --> tests/ui/question-mark-await-type-error.rs:28:9
    |
28  |         async {}.await?; //~ ERROR the `?` operator can only be applied to values that implement `std::ops::Try`
    |         ^^^^^^^^^^^^^^^ the `?` operator cannot be applied to type `()`
    |
    = help: the trait `Try` is not implemented for `()`
note: required by `branch`
   --> $RUST/core/src/ops/try_trait.rs
    |
    |     fn branch(self) -> ControlFlow<Self::Residual, Self::Output>;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the `?` operator can only be used in an async function that returns `Result` or `Option` (or another type that implements `FromResidual`)
   --> tests/ui/question-mark-await-type-error.rs:28:23
    |
25  |   async fn async_fn_and_for_await() {
    |  ___________________________________-
26  | |     #[for_await]
27  | |     for _i in stream(2) {
28  | |         async {}.await?; //~ ERROR the `?` operator can only be applied to values that implement `std::ops::Try`
    | |                       ^ cannot use the `?` operator in an async function that returns `()`
29  | |     }
30  | | }
    | |_- this function should return `Result` or `Option` to accept `?`
    |
    = help: the trait `FromResidual<_>` is not implemented for `()`
note: required by `from_residual`
   --> $RUST/core/src/ops/try_trait.rs
    |
    |     fn from_residual(residual: R) -> Self;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the `?` operator can only be applied to values that implement `Try`
   --> tests/ui/question-mark-await-type-error.rs:36:18
    |
36  |         async {}.await?; //~ ERROR the `?` operator can only be applied to values that implement `std::ops::Try`
    |                  ^^^^^^ the `?` operator cannot be applied to type `()`
    |
    = help: the trait `Try` is not implemented for `()`
note: required by `branch`
   --> $RUST/core/src/ops/try_trait.rs
    |
    |     fn branch(self) -> ControlFlow<Self::Residual, Self::Output>;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the `?` operator can only be used in a generator that returns `Result` or `Option` (or another type that implements `FromResidual`)
   --> tests/ui/question-mark-await-type-error.rs:36:23
    |
32  | #[stream(item = i32)]
    | --------------------- this function should return `Result` or `Option` to accept `?`
...
36  |         async {}.await?; //~ ERROR the `?` operator can only be applied to values that implement `std::ops::Try`
    |                       ^ cannot use the `?` operator in a generator that returns `()`
    |
    = help: the trait `FromResidual<_>` is not implemented for `()`
note: required by `from_residual`
   --> $RUST/core/src/ops/try_trait.rs
    |
    |     fn from_residual(residual: R) -> Self;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
